### ISSUE TITLE
Most appropriate way to translate references

### ISSUE THREAD

**Issue by juodaspaulius** [juodaspaulius/clafer#19]
Let's say we have a reference `c1` in a model:

```
abstract B
final A
  d : B -> D
D
```

Every reference is reified, so Alloy generator creates an additional signature `sig c0_d` to encode the reference. In case the reference is not final, we need to attach State set to the reference relation. A small dilemma is whether to attach it to relation in reified class or in the parent class?..  So AlloyLtl generator can translate above into:

Case 1:

```
sig c0_A
{ r_c0_d : c0_d -> State } {...}

sig c0_d
{ c0_d_ref : c0_D } {...}
```

or Case 2:

```
sig c0_A
{ r_c0_d : c0_d } {...}

sig c0_d
{ c0_d_ref : c0_D -> State } {...}
```

Any input on this?

**Comment by mantkiew** [comment]
Or, actually to both! We have Case 3:

```
sig c0_A
{ r_c0_d : c0_d -> State } {...}

sig c0_d
{ c0_d_ref : c0_D -> State } {...}
```

The keyword `final` refers to the clafer and, for reference clafers, also to the ref relationship. So, for a model 

```
final A
  final d : B -> D
```

we should have static clafer

```
sig c0_A
{ r_c0_d : c0_d } {...}

sig c0_d
{ c0_d_ref : c0_D } {...}
```

However, when we don't have `final d`, then both the reference object  (instance of `d`) and the reference value (tuple in the relation `c0_d_ref`) can change. So, both should have the `-> State`, in my opinion (Case 3).

**Comment by wasowski** [comment]
I believe that Michał is write.  We can have a case that the pointed to Clafer instance disappears at some point, but the reference is redirected to another instance of the same type.

**Comment by mantkiew** [comment]
Some ideas from the discussion:

final dr -> D   // makes all final
finalref dr -> D   // makes the dr final but Ds can change
finaltarget dr -> D   //forces dr to change to a different instance if D changes

**Comment by juodaspaulius** [comment]
Yes, using option 3 adds more possibilities, which can be restricted to option 1/2 using modifiers. Nice idea, although it might add complexity for the end user. 

I will implement it, so we can try it it practice.
